ryan_savill/03_background_subtraction/readme.md

```diff
@@ -0,0 +1,281 @@
+
+# subtraction from the original to obtain the tophat filter
+tophat_trib = tribolium-max_of_min_trib
+
+fig, (ax1, ax2, ax3, ax4) = plt.subplots(1, 4, figsize =(20,20))
+ax1.imshow(tribolium, cmap = 'gray')
+ax1.set_title('Original')
+ax2.imshow(minimum_trib, cmap = 'gray')
+ax2.set_title('Minimum')
+ax3.imshow(max_of_min_trib, cmap = 'gray')
+ax3.set_title('Max of Min')
+ax4.imshow(tophat_trib, cmap = 'gray')
+ax4.set_title('Tophat')
+```
+
+
+    
+![png](output_5_1.png)
+    
+
+
+This in essence is the tophat-filter and we could now put it into a function to use to make our life easier or we can plot it with different values for the max and min filters to see what effect that has on the background subtraction:
+
+
+```python
+def diytophat(image, minsize=25, maxsize=25):
+    from scipy.ndimage import maximum_filter, minimum_filter
+    minimum = minimum_filter(image,minsize)
+    max_of_min = maximum_filter(minimum, maxsize)
+    tophat = image - max_of_min
+    return tophat
+
+# creating a range of filtersizes to iterate through (from 5 to 105 in steps of size 20)
+filtersizes = range(5,105,20)
+
+# initialising the plot
+fig, axs = plt.subplots(len(filtersizes), len(filtersizes), figsize=(20,20))
+
+# iterating over the filtersizes for min and max
+for index_1, i in enumerate(filtersizes):
+    for index_2, j in enumerate(filtersizes):
+        # making the image and showing it on the plot 
+        axs[index_1,index_2].imshow(diytophat(tribolium,i,j),cmap='gray')
+        axs[index_1,index_2].set_title("min = {minval}, max = {maxval}".format(minval = i, maxval = j))
+plt.show()
+```
+
+
+    
+![png](output_7_0.png)
+    
+
+
+From these plots we can see that keeping the filtersizes the same yields the best results, but having the maximum filter slightly smaller than the minimum filter also produces plausible results. Any large differences between the values seem to distort the image substancially. Since equal values seem to work the best we will modify our function to only have one filter size as input:
+
+
+```python
+def diytophat(image, size=25):
+    from scipy.ndimage import maximum_filter, minimum_filter
+    minimum = minimum_filter(image,size)
+    max_of_min = maximum_filter(minimum,size)
+    tophat = image - max_of_min
+    return tophat
+```
+
+One of the main benefits of python is that you don't need to write a lot of code if you are doing things that are common, which is why to apply a tophat filter we can also just use scikit image to apply a tophat filter! The downside is that there are usually some peculiarities with the prepackaged code. In scikit image for some filters we need a structuring element. Basically what this is is a 2d (or 3d if we are processing 3d images) matrix filled with 1 where the filter should be applied and 0 otherwise. The filters need this as a parameter to know how to apply the filter. Now let's generate one to see what it looks like:
```

Comment:
```suggestion
One of the main benefits of python is that you don't need to write a lot of code if you are doing things that are common, which is why to apply a top-hat filter we can also just use [scikit image](https://scikit-image.org/) to apply a top-hat filter! The downside is that there are usually some peculiarities with the prepackaged code. In scikit image for some filters we need a [structuring element](https://homepages.inf.ed.ac.uk/rbf/HIPR2/strctel.htm). Basically, this is a 2d (or 3d if we are processing 3d images) matrix filled with 1 where the filter should take a pixel into account and 0 otherwise. The filters need this as a parameter to know how to apply the filter spatially. Now let's generate one to see what it looks like:
```
Added links